Detect multiple crate versions on method not found

estebank opened the pull request:
When a type comes indirectly from one crate version but the imported trait comes from a separate crate version, the called method won't be found. We now show additional context:

```
error[E0599]: no method named `foo` found for struct `dep_2_reexport::Type` in the current scope
 --> multiple-dep-versions.rs:8:10
  |
8 |     Type.foo();
  |          ^^^ method not found in `Type`
  |
note: there are multiple different versions of crate `dependency` in the dependency graph
 --> multiple-dep-versions.rs:4:32
  |
4 | use dependency::{do_something, Trait};
  |                                ^^^^^ `dependency` imported here doesn't correspond to the right crate version
  |
 ::: ~/rust/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:4:1
  |
4 | pub trait Trait {
  | --------------- this is the trait that was imported
  |
 ::: ~/rust/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-2.rs:4:1
  |
4 | pub trait Trait {
  | --------------- this is the trait that is needed
5 |     fn foo(&self);
  |        --- the method is available for `dep_2_reexport::Type` here
```

Fix #128569, fix #110926, fix #109161, fix #81659, fix #51458, fix #32611. Follow up to #124944.

estebank committed:
Detect multiple crate versions on method not found

When a type comes indirectly from one crate version but the imported trait comes from a separate crate version, the called method won't be found. We now show additional context:

```
error[E0599]: no method named `foo` found for struct `dep_2_reexport::Type` in the current scope
 --> multiple-dep-versions.rs:8:10
  |
8 |     Type.foo();
  |          ^^^ method not found in `Type`
  |
note: you have multiple different versions of crate `dependency` in your dependency graph
 --> multiple-dep-versions.rs:4:32
  |
4 | use dependency::{do_something, Trait};
  |                                ^^^^^ `dependency` imported here doesn't correspond to the right crate version
  |
 ::: ~/rust/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:4:1
  |
4 | pub trait Trait {
  | --------------- this is the trait that was imported
  |
 ::: ~/rust/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-2.rs:4:1
  |
4 | pub trait Trait {
  | --------------- this is the trait that is needed
5 |     fn foo(&self);
  |        --- the method is available for `dep_2_reexport::Type` here
```

compiler-errors commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -4041,6 +4050,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                let name = self.tcx.crate_name(trait_candidate.def_id.krate);
+                if trait_candidate.def_id.krate != item.def_id.krate && name == pick_name {
+                    let msg = format!(
+                        "you have multiple different versions of crate `{name}` in your \
```
I feel like saying "you" here makes this read weirdly

```suggestion
                        "there have multiple different versions of crate `{name}` in the \
```

estebank replied:
There are a couple of other strings in this module that use `you`, we should clean those up too, separately.

estebank committed:
reword message

estebank commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -4041,6 +4050,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        rcvr_ty: Ty<'_>,
+    ) -> bool {
+        let pick_name = self.tcx.crate_name(item.def_id.krate);
+        if let Some(map) = self.tcx.in_scope_traits_map(owner) {
+            for trait_candidate in map.to_sorted_stable_ord().into_iter().flat_map(|v| v.1.iter()) {
```
This is the thing I'm slightly uneasy about: for *every* error of a method not found that triggers the prior conditions, we'll be doing a linear scan of all traits in scope. Don't have a good sense of whether that will be a significant slow-down for compiles with errors (not a good way of testing that difference).

fee1-dead replied:
Please take some time to fuzz this. Would it be possible to trigger this in some normal session by removing some imports or upgrading a crate to a newer version (where the trait method is removed in the newer version and the older version is still imported by some transitive dependencies)?

It's very possible that some large project trying to refactor something will hit this if this is a significant slowdown.

estebank replied:
I think putting this behind a query that allocates a `HashMap<crate name, Vec<trait DefId>>` and returns it would reduce the amortized cost significantly and eliminate the risk of causing trouble.

estebank replied:
Ok, I changed it to use `in_scope_traits` instead of `in_scope_traits_map`, which will be much cheaper (at the cost of only looking at traits in the current scope, which is fine).

estebank replied:
@fee1-dead given the current approach (where we're looking at in-scope traits for a *specific* scope), which is something we're already doing for every method resolution, I feel more comfortable with it. I'm also not quite sure how to synthesize a big enough corpus of "wrong crate" to properly stress this.

estebank committed:
Account for fully-qualified path case of conflicting crate versions

When encountering the following, mention the precense of conflicting crates:

```
error[E0599]: no function or associated item named `get_decoded` found for struct `HpkeConfig` in the current scope
   --> src/main.rs:7:17
    |
7   |     HpkeConfig::get_decoded(&foo);
    |                 ^^^^^^^^^^^ function or associated item not found in `HpkeConfig`
    |
note: if you're trying to build a new `HpkeConfig`, consider using `HpkeConfig::new` which returns `HpkeConfig`
   --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/janus_messages-0.3.1/src/lib.rs:908:5
    |
908 | /     pub fn new(
909 | |         id: HpkeConfigId,
910 | |         kem_id: HpkeKemId,
911 | |         kdf_id: HpkeKdfId,
912 | |         aead_id: HpkeAeadId,
913 | |         public_key: HpkePublicKey,
914 | |     ) -> HpkeConfig {
    | |___________________^
note: there are multiple different versions of crate `prio` in the dependency graph
   --> src/main.rs:1:5
    |
1   | use prio::codec::Decode;
    |     ^^^^^^^^^^^^^^^^^^^ `prio` imported here doesn't correspond to the right crate version
    |
   ::: ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/prio-0.9.1/src/codec.rs:35:1
    |
35  | pub trait Decode: Sized {
    | ----------------------- this is the trait that was imported
    |
   ::: ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/prio-0.10.3/src/codec.rs:35:1
    |
35  | pub trait Decode: Sized {
    | ----------------------- this is the trait that is needed
...
43  |     fn get_decoded(bytes: &[u8]) -> Result<Self, CodecError> {
    |     -------------------------------------------------------- the method is available for `HpkeConfig` here
help: there is an associated function `decode` with a similar name
    |
7   |     HpkeConfig::decode(&foo);
    |                 ~~~~~~
```

estebank committed:
Ignore auto-deref for multiple crate version note

As per the case presented in #128569, we should be showing the extra info even if auto-deref is involved.

estebank commented:
I think after landing this PR the only error that still needs to check for multiple crate versions is E0277. E0308 already has a check (even though it isn't great), and this addresses E0599 for every case I could get my hands on.

fee1-dead commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -4041,6 +4075,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        item_span: Span,
+        owner: hir::OwnerId,
+        rcvr_ty: Ty<'_>,
+    ) -> bool {
```
The bool returned here is very unconventional. Normally, we return `false` if we did not emit something and vice versa. this is a `detect_*` method, so I'd expect the boolean to be flipped. I'd recommend renaming the `suggest` var to something like `trait_in_other_version_found`, and making the return value for this correspond to the convention.

fee1-dead commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -3676,7 +3686,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                .filter_by_name_unhygienic(item_name.name)
+                .next()
+        {
+            // See if the `Type::function(val)` where `function` wasn't found corresponds to a
+            // `Trait` that is imported directly, but `Type` came from a different version of the
+            // same crate.
```
Please also add a test for this.

estebank committed:
Rework suggestion method

Make checking slightly cheaper (by restricting to the right item only).

Add tests.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10327508554/job/28592823970) [(plain)](https://github.com/rust-lang/rust/commit/882f4a46ed4d687cab9be75463ca14baa0a44e23/checks/28592823970/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.828 Building wheels for collected packages: reuse
#13 2.829   Building wheel for reuse (pyproject.toml): started
#13 3.073   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.074   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.075   Stored in directory: /tmp/pip-ephem-wheel-cache-qweolzk_/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.077 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.473 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.474 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.014 Collecting virtualenv
#13 4.014 Collecting virtualenv
#13 4.062   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.289      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 25.2 MB/s eta 0:00:00
#13 4.350 Collecting filelock<4,>=3.12.2
#13 4.357   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.378 Collecting distlib<1,>=0.3.7
#13 4.385   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.397      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 48.0 MB/s eta 0:00:00
#13 4.430 Collecting platformdirs<5,>=3.9.1
#13 4.438   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.520 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.710 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      221120 kB
DirectMap2M:     7118848 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/ca5d25e2c41f5a6b4ce65c681bf2f94c7ead1f14/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-ca5d25e2c41f5a6b4ce65c681bf2f94c7ead1f14-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5460 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:21: line longer than 100 chars
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:33: line longer than 100 chars
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:35: line longer than 100 chars
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:61: line longer than 100 chars
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:66: line longer than 100 chars
##[error]tidy error: /checkout/tests/run-make/crate-loading/rmake.rs:83: line longer than 100 chars
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 42.7 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-24.2
linting python files
All checks passed!
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Aug 10 00:45:12 UTC 2024
  network time: Sat, 10 Aug 2024 00:45:12 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10327640120/job/28593154994) [(plain)](https://github.com/rust-lang/rust/commit/515a88b72a80e7109fef9e709926e0bda4908fb5/checks/28593154994/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.77s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/crate-loading stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out" && env -u RUSTFLAGS AR="ar" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-17/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-17/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake"
--- stderr -------------------------------
--- stderr -------------------------------
=== HAYSTACK ===
error[E0277]: the trait bound `dep_2_reexport::Type: Trait` is not satisfied
##[error]  --> multiple-dep-versions.rs:7:18
7  |     do_something(Type);
   |     ------------ ^^^^ the trait `Trait` is not implemented for `dep_2_reexport::Type`
   |     |
   |     required by a bound introduced by this call
   |     required by a bound introduced by this call
   |
help: there are multiple different versions of crate `dependency` the your dependency graph
  --> multiple-dep-versions.rs:1:1
1  | extern crate dep_2_reexport;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one version of crate `dependency` is used here, as a dependency of crate `foo`
2  | extern crate dependency;
   | ^^^^^^^^^^^^^^^^^^^^^^^^ one version of crate `dependency` is used here, as a direct dependency of the current crate
   | ^^^^^^^^^^^^^^^^^^^^^^^^ one version of crate `dependency` is used here, as a direct dependency of the current crate
note: two types coming from two different versions of the same crate are different types even if they look the same
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:3:1
3  | pub struct Type(pub i32);
   | ^^^^^^^^^^^^^^^ this type implements the required trait
4  | pub trait Trait {
   | --------------- this is the required trait
   | --------------- this is the required trait
   |
  ::: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-2.rs:3:1
3  | pub struct Type;
   | ^^^^^^^^^^^^^^^ this type doesn't implement the required trait
   = help: you can use `cargo tree` to explore your dependency tree
note: required by a bound in `do_something`
note: required by a bound in `do_something`
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:12:24
   |
12 | pub fn do_something<X: Trait>(_: X) {}
   |                        ^^^^^ required by this bound in `do_something`

error[E0599]: no method named `foo` found for struct `dep_2_reexport::Type` in the current scope
##[error] --> multiple-dep-versions.rs:8:10
8 |     Type.foo();
  |          ^^^ method not found in `Type`
  |
  |
note: there are multiple different versions of crate `dependency` in the dependency graph
 --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-2.rs:4:1
4 | pub trait Trait {
  | ^^^^^^^^^^^^^^^ this is the trait that is needed
5 |     fn foo(&self);
5 |     fn foo(&self);
  |     -------------- the method is available for `dep_2_reexport::Type` here
 ::: multiple-dep-versions.rs:4:32
  |
4 | use dependency::{do_something, Trait};
4 | use dependency::{do_something, Trait};
  |                                ----- `Trait` imported here doesn't correspond to the right version of crate `dependency`
  |
 ::: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:4:1
4 | pub trait Trait {
  | --------------- this is the trait that was imported


error[E0599]: no function or associated item named `bar` found for struct `dep_2_reexport::Type` in the current scope
##[error] --> multiple-dep-versions.rs:9:11
9 |     Type::bar();
  |           ^^^ function or associated item not found in `Type`
  |
  |
note: there are multiple different versions of crate `dependency` in the dependency graph
 --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-2.rs:4:1
4 | pub trait Trait {
  | ^^^^^^^^^^^^^^^ this is the trait that is needed
5 |     fn foo(&self);
6 |     fn bar();
6 |     fn bar();
  |     --------- the method is available for `dep_2_reexport::Type` here
 ::: multiple-dep-versions.rs:4:32
  |
4 | use dependency::{do_something, Trait};
4 | use dependency::{do_something, Trait};
  |                                ----- `Trait` imported here doesn't correspond to the right version of crate `dependency`
  |
 ::: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/crate-loading/rmake_out/multiple-dep-versions-1.rs:4:1
4 | pub trait Trait {
  | --------------- this is the trait that was imported

warning: unused import: `Trait`
---

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.

=== NEEDLE ===
error[E0277]: the trait bound `dep_2_reexport::Type: Trait` is \
  --> multiple-dep-versions.rs:7:18
   |
7  |     do_something(Type);
   |     ------------ ^^^^ the trait `Trait` is not implemented for `dep_2_reexport::Type`
   |     ------------ ^^^^ the trait `Trait` is not implemented for `dep_2_reexport::Type`
   |     |
   |     required by a bound introduced by this call
   |
help: there are multiple different versions of crate `dependency` the your dependency graph
  --> multiple-dep-versions.rs:1:1
1  | extern crate dep_2_reexport;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one version of crate `dependency` is used here, as a dependency \
of crate `foo`
2  | extern crate dependency;
```

</details>

estebank committed:
Properly differentiate between methods and assoc fns

fee1-dead reviewed (approved):
Thanks!

fee1-dead commented:
@bors r+

bors commented:
:pushpin: Commit 110b19b2b6a9896d4d73b8bd8b735fa9c2503d1e has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 110b19b2b6a9896d4d73b8bd8b735fa9c2503d1e -->
<!-- homu: {"type":"Approved","sha":"110b19b2b6a9896d4d73b8bd8b735fa9c2503d1e","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 110b19b2b6a9896d4d73b8bd8b735fa9c2503d1e with merge 9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1...
<!-- homu: {"type":"BuildStarted","head_sha":"110b19b2b6a9896d4d73b8bd8b735fa9c2503d1e","merge_sha":"9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10433064496/job/28896144863)
Approved by: fee1-dead
Pushing 9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"fee1-dead","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10433064496/job/28896144863"},"merge_sha":"9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1"} -->

rust-timer commented:
Finished benchmarking commit (9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1): [comparison URL](https://perf.rust-lang.org/compare.html?start=c6f81a452e291dd4ff5bcbc6ecb9cd6f2271f164&end=9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c6f81a452e291dd4ff5bcbc6ecb9cd6f2271f164&end=9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c6f81a452e291dd4ff5bcbc6ecb9cd6f2271f164&end=9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1&stat=max-rss)
<details>
<summary>Results (primary 2.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1% | [2.1%, 2.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.1% | [2.1%, 2.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c6f81a452e291dd4ff5bcbc6ecb9cd6f2271f164&end=9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=c6f81a452e291dd4ff5bcbc6ecb9cd6f2271f164&end=9b318d2e93ce35e7ba32d8cfa96a1dbe63a7bed1&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 749.433s -> 750.64s (0.16%)
**Artifact size**: 339.08 MiB -> 339.13 MiB (0.01%)
